types/react-query/index.d.ts

```diff
@@ -1,142 +1,420 @@
+    variables?: TVariables;
+    queryFn: QueryFunctionWithVariables<TResult, TKey, TVariables>;
+    config?: QueryOptions<TResult>;
+}): PaginatedQueryResult<TResult>;
+
+// useInfiniteQuery
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, TKey, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, [TKey], TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, [TKey], TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+export function useInfiniteQuery<
+    TResult,
+    TKey extends AnyQueryKey,
+    TMoreVariable,
+    TVariables extends AnyVariables = []
+>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>;
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>;
+}): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export interface QueryOptions<TResult> {
+export type QueryKeyPart = string | object | boolean | number | null | readonly QueryKeyPart[] | null | undefined;
+export type AnyQueryKey = readonly [string, ...QueryKeyPart[]]; // this forces the key to be inferred as a tuple
+export type AnyVariables = readonly [] | readonly [any, ...any[]]; // this forces the variables to be inferred as a tuple
+
+export type QueryFunction<TResult, TKey extends AnyQueryKey> = (...key: TKey) => Promise<TResult>;
```

Comment:
Hi @Igorbek 
Thanks for all the work you've done.
Should we include/allow `string` queryKeys here? The library seems to [allow it](https://github.com/tannerlinsley/react-query#string-only-query-keys)

Reply:
You're welcome.
It is allowed, but not in the query function (the one that fetches data), it is allowed in `useQuery` and friends. Query functions, on the other hand, always accept the keys as set of arguments, no matter is that single string or an array.